### ISSUE TITLE
Middleware to cache the body for multiple passes

### ISSUE THREAD

**Issue by rossabaker** [http4s/http4s#5199]
This keeps coming up. (e.g., #5190).  I don't recall that we've ever written a middleware that caches all the bytes and passes a re-readable body downstream.  If we haven't, let's stop handwaving at it and provide it.

**Comment by kryptt** [comment]
it would be nice if this middleware is also aware of http cache headers...

**Comment by bplommer** [comment]
> it would be nice if this middleware is also aware of http cache headers...

I don't think that's applicable here? This is memoizing an individual message body as an implementation detail for handling it correctly, not caching in the http sense.

**Comment by kryptt** [comment]
> 
> 
> > it would be nice if this middleware is also aware of http cache headers...
> 
> I don't think that's applicable here? This is memoizing an individual message body as an implementation detail for handling it correctly, not caching in the http sense.

Still, knowledge of the fact a response hasn't changed (because it has been cached / memoized) is something the transport layer (HTTP) is able to share.... 

I agree there are situations where you still want the client to revalidate / query again, etc.. but curiously enough that is also in the HTTP cache spec.... :D

**Comment by FunFunFine** [comment]
Perhaps it's possible to reuse `ContextRequest` here like with `AuthedRequest`:
```scala
type CachedRequest[F[_]] = ContextRequest[F, Array[Byte]]
```
This way the user still has the ability to compile the body again, but at the same can use the cached one.

**Comment by rossabaker** [comment]
A second compilation is not going to work for a socket-backed stream, and whoever goes second might not know that they're going second.  I think we want to make it transparent.

**Comment by FunFunFine** [comment]
Well, I did some thinking and came to this dead simple solution for the server caching middleware. It literally fits in one snippet:

```scala
import cats.data.Kleisli
import cats.effect.Concurrent
import fs2.Stream
import org.http4s.HttpApp
import cats.implicits._


  def cachingMiddleware[F[_]: Concurrent]: HttpApp[F] => HttpApp[F] =
    app =>
      Kleisli {
        req =>
          for {
            body <- req.body.compile.toList
            cachedReq =
              req.withBodyStream(Stream.evalSeq[F, List, Byte](body.pure[F]))
            resp <- app(cachedReq)
          } yield resp
      }
```
Of course, it needs some work done yet (docs, performance, etc.) but I think you get the idea. Would it be sufficient? If so, then I could make a PR.

**Comment by kryptt** [comment]
I was playing with a transparent caching fs2 pipe as well.... 

https://gist.github.com/kryptt/e20e861df414a643503112b408abde38

**Comment by armanbilge** [comment]
@FunFunFine @kryptt so, who's going to PR :)

**Comment by rossabaker** [comment]
Nothing about @kryptt's looks specific to HTTP.  Something around it that understood HTTP caching directives would make it a better candidate for http4s, but the current gist might be more appropriate for something upstream.

@FunFunFine's integrates with the request and would definitely be in scope for http4s.  I have some feedback on it that's probably easier done on a live PR.

**Comment by kryptt** [comment]
Yes, what I was doing was a general caching of fs2 streams... unfortunately scalacache is having a hard time getting any updates it seems...
